Add faro.receiver component (#5314)
grafana/agent: docs/sources/flow/reference/components/faro.receiver.md

---
aliases:
- /docs/grafana-cloud/agent/flow/reference/components/faro.receiver/
- /docs/grafana-cloud/monitor-infrastructure/agent/flow/reference/components/faro.receiver/
- /docs/grafana-cloud/monitor-infrastructure/integrations/agent/flow/reference/components/faro.receiver/
canonical: https://grafana.com/docs/agent/latest/flow/reference/components/faro.receiver/

> title: faro.receiver

Reviewer: ```suggestion
title: faro.receiver
description: Learn about the faro.receiver
```

We need to add the description: metadata in all files. It doesn't have to be elaborate, just present with some simple text.

---

# faro.receiver

`faro.receiver` accepts web application telemetry data from the [Grafana Faro Web SDK][faro-sdk]
and forwards it to other components for future processing.

[faro-sdk]: https://github.com/grafana/faro-web-sdk

## Usage

```river
faro.receiver "LABEL" {
    output {

>         logs   = [LOGS_RECEIVERS]

Reviewer 2: Is this an OTel log receiver (like traces below), or a loki.* component?

*edit*: I see this is explained in the `output` block below, nvm.

server > rate_limiting | [rate_limiting][] | Configures rate limiting for the HTTP server. | no
sourcemaps | [sourcemaps][] | Configures sourcemap retrieval. | no
sourcemaps > location | [location][] | Configures on-disk location for sourcemap retrieval. | no
output | [output][] | Configures where to send collected telemetry data. | yes

[server]: #server-block
[rate_limiting]: #rate_limiting-block
[sourcemaps]: #sourcemaps-block
[location]: #location-block
[output]: #output-block

### server block

The `server` block configures the HTTP server managed by the `faro.receiver`
component. Clients using the [Grafana Faro Web SDK][faro-sdk] forward telemetry
data to this HTTP server for processing.

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`listen_address` | `string` | Address to listen for HTTP traffic on. | `127.0.0.1` | no
`listen_port` | `number` | Port to listen for HTTP traffic on. | `12347` | no
`cors_allowed_origins` | `list(string)` | Origins for which cross-origin requests are permitted. | `[]` | no
`api_key` | `secret` | Optional API key to validate client requests with. | `""` | no

> `max_allowed_payload_size` | `number` | Maximum size (in bytes) for client requests. | `5e6` | no

Reviewer 2: nit: In the future we can have this use unit2bytes like we do for other components. Let's just leave a TODO comment for it.

Author: I'll change that now; shouldn't be too much effort.

By default, telemetry data is only accepted from applications on the same local
network as the browser. To accept telemetry data from a wider set of clients,
modify the `listen_address` attribute to the IP address of the appropriate
network interface to use.

> The `cors_allowed_origins` argument determines what oriigns browser requests

Reviewer: ```suggestion
The `cors_allowed_origins` argument determines what origins browser requests
```

may come from. The default value, `[]`, disables CORS support. To support
requests from all origins, set `cors_allowed_origins` to `["*"]`. The `*`
character indicates a wildcard.

When the `api_key` argument is non-empty, client requests must have an HTTP
header called `X-API-Key` matching the value of the `api_key` argument.
Requests that are missing the header or have the wrong value are rejected with
an `HTTP 401 Unauthorized` status code. If the `api_key` argument is empty, no
authentication checks are performed, and the `X-API-Key` HTTP header is
ignored.

### rate_limiting block

The `rate_limiting` block configures rate limiting for client requests.

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`enabled` | `bool` | Whether to enable rate limiting. | `true` | no
`rate` | `number` | Rate of allowed requests per second. | `50` | no
`burst_size` | `number` | Allowed burst size of requests. | `100` | no

Rate limiting functions as a [token bucket algorithm][token-bucket], where

> a bucket has maximum capacity for up to `burst_size` requests, and refills at a

Reviewer: ```suggestion
a bucket has a maximum capacity for up to `burst_size` requests and refills at a
```

rate of `rate` per second.

Each HTTP request drains the capacity of the bucket by one. Once the bucket is

> empty, HTTP requests are rejected with a `HTTP 429 Too Many Requests` status

Reviewer: ```suggestion
empty, HTTP requests are rejected with an `HTTP 429 Too Many Requests` status
```

> code until the bucket has more availble capacity.

Reviewer: ```suggestion
code until the bucket has more available capacity.
```

Configuring the `rate` argument determines how fast the bucket refills, and
configuring the `burst_size` argument determines how many requests can be
received in a burst before the bucket is empty and starts rejecting requests.

[token-bucket]: https://en.wikipedia.org/wiki/Token_bucket

### sourcemaps block

The `sourcemaps` block configures how to retrieve sourcemaps. Sourcemaps are
then used to transform file and line information from minified code into the
file and line information from the original source code.

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`download` | `bool` | Whether to download sourcemaps. | `true` | no
`download_from_origins` | `list(string)` | Which origins to download sourcemaps from. | `["*"]` | no
`download_timeout` | `duration` | Timeout when downloading sourcemaps. | `"1s"` | no

When exceptions are sent to the `faro.receiver` component, it can download

> sourcemaps from the web application. This behavior can be disabled by setting

Reviewer: ```suggestion
sourcemaps from the web application. You can disable this behavior by setting
```

The `minified_path_prefix` argument determines the prefix of paths to
Javascript files, such as `http://example.com/`. The `path` argument then
determines where to find the sourcemap for the file.

For example, given the following location block:

```
location {
    path                 = "/var/my-app/build"
    minified_path_prefix = "http://example.com/"
}
```

To look up the sourcemaps for a file hosted at `http://example.com/foo.js`, the
`faro.receiver` component will:

1. Remove the minified path prefix to extract the path to the file (`foo.js`).
2. Search for that file path with a `.map` extension (`foo.js.map`) in `path`
   (`/var/my-app/build/foo.js.map`).

Optionally, the value for the `path` argument may contain `{{ .Release }}` as a
template value, such as `/var/my-app/{{ .Release }}/build`. The template value
will be replaced with the release value provided by the [Faro Web App SDK][faro-sdk].

### output block

The `output` block specifies where to forward collected logs and traces.

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------

> `logs` | `list(LogsReceiver)` | A list of `loki` components to forward lgos to. | `[]` | no

Reviewer 2: ```suggestion
`logs` | `list(LogsReceiver)` | A list of `loki` components to forward logs to. | `[]` | no
```

* `faro_receiver_inflight_requests` (gauge): Current number of inflight requests.
* `faro_receiver_sourcemap_cache_size` (counter): Number of items in sourcemap cache per origin.
* `faro_receiver_sourcemap_downloads_total` (counter): Total number of sourcemap downloads performed per origin and status.
* `faro_receiver_sourcemap_file_reads_total` (counter): Total number of sourcemap retrievals using the filesystem per origin and status.

## Example

```river
faro.receiver "default" {
    server {
        listen_address = "NETWORK_ADDRESS"
    }

    sourcemaps {
        location {
            path                 = "PATH_TO_SOURCEMAPS"
            minified_path_prefix = "WEB_APP_PREFIX"
        }
    }

    output {
        logs   = [loki.write.default.receiver]
        traces = [otelcol.exporter.otlp.traces.input]
    }
}

loki.write "default" {
    endpoint {
        url = "https://LOKI_ADDRESS/api/v1/push"
    }
}

otelcol.exporter.otlp "traces" {
    client {
        endpoint = "OTLP_ADDRESS"
    }
}
```

Replace the following:

* `NETWORK_ADDRESS`: IP address of the network interface to listen to traffic

>   on. This IP adddress must be reachable by browsers using the web application

Reviewer: ```suggestion
  on. This IP address must be reachable by browsers using the web application
```
